From patchwork Fri Oct 12 18:04:28 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10639081
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id A5A4E15E2
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 12 Oct 2018 18:04:43 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 977B72C276
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 12 Oct 2018 18:04:43 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 88A942C294; Fri, 12 Oct 2018 18:04:43 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 3AE3E2C276
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 12 Oct 2018 18:04:43 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726663AbeJMBiX (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 12 Oct 2018 21:38:23 -0400
Received: from mail-io1-f73.google.com ([209.85.166.73]:39079 "EHLO
        mail-io1-f73.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1725929AbeJMBiW (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 12 Oct 2018 21:38:22 -0400
Received: by mail-io1-f73.google.com with SMTP id x5-v6so11888197ioa.6
        for <kvm@vger.kernel.org>; Fri, 12 Oct 2018 11:04:41 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=eqbLe1CIeIQoXM0Z71Gg63vrSkfd/iUjX7ZFIx68600=;
        b=EgQLr0Jj7v5R7Yuw9a3fxCs+uLTLO+lQ2kXpJQjZDgFFHxVTtoOBY+YK9Qx0rw/3Va
         fD0yqp78K2YyItsXhu5vzkwcvMS1X8SgG/aMHPhIK85ttzgJy/nkVQoOJMgW/Ktf3zLb
         DrtpYzXE7HzbNpP4odFcTZxUZiT4jGtPZIJVVB3nohUeNyVh/dnDGs50AZQ3Ta9JtXfT
         ZRTpNYPqBhnNapRTWNKUy+NEFFbfqqm31np332DnibDBpWzzaz+o+TyiM5mxDMvy3oqg
         rSS/yniGmrTpJNv/+VyZgn/8OkRMclMcxPsCPVsw5e2uestlwDg+hoePtCpRE+V7fLV1
         jo8w==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=eqbLe1CIeIQoXM0Z71Gg63vrSkfd/iUjX7ZFIx68600=;
        b=dpY0YltRV17+MUh6KB4gzqODcmp3mX65m0zOWA6lxr8Xu1kCOSyuPlaU5h1j3FywR0
         5zpsapo/8wM84/Mr1CgWC6qvi7Mf5ab49+aJCe916gdI9Sms83w+Tl8yRphy7fbNslR9
         JjtBrSHzjctaSu2cn4F/34Qzs488ckmzElb+A50+V8jdn3st1JTHyKvb4N/gjmH9iz1L
         ihjEQtDWRNgHGFiLTRmrEmvAOHBTcly6q2+fTOX0Sw0KD4YAF6asBt0wy80IoSeuEUx0
         jKS9FnHISBcFea492hLny+KuQvnEAoIn+5ZIhz5LQ6OKkN0FqpmoMDtR0h7Sgn5vsukd
         Fm4A==
X-Gm-Message-State: ABuFfojXHOIVEfe3kGx3KBIDYHcL4URTprMLRIaj5OWXX1kmouwJvhiX
        ppetvxBcRMLZZlUcfRwRIXmR74VVZNaZcMRTxVzOoKegm0tNpeyniC4s8gblXED48vEE+HgD7FX
        uqyd0e0im3QQE9yBeg57eFHMmWpD+xs7Lq24QRL3yPMvxeiqKLnJ40n8Otu7VHZQ=
X-Google-Smtp-Source: 
 ACcGV61cSKlTwqQRNz7PJqSYBTocbxPFrm1avX4xmmRN1KaHdhaG2X5zITBM/yQiSIEcfXYFjgjIIQz/7+c9ng==
X-Received: by 2002:a24:9714:: with SMTP id
 k20-v6mr5911793ite.0.1539367480597;
 Fri, 12 Oct 2018 11:04:40 -0700 (PDT)
Date: Fri, 12 Oct 2018 11:04:28 -0700
Message-Id: <20181012180429.259916-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.19.1.331.ge82ca0e54c-goog
Subject: [PATCH 1/2] kvm: x86: Bounds-check argument to
 x86_exception_has_error_code
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Marc Orr <marcorr@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

x86_exception_has_error_code should deterministically return false if
its operand is greater than or equal to 32.

Fixes: 0447378a4a793 ("kvm: vmx: Nested VM-entry prereqs for event inj.")
Cc: Marc Orr <marcorr@google.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 arch/x86/kvm/x86.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/kvm/x86.h b/arch/x86/kvm/x86.h
index 67b9568613f3..b7890b248e64 100644
--- a/arch/x86/kvm/x86.h
+++ b/arch/x86/kvm/x86.h
@@ -116,7 +116,7 @@ static inline bool x86_exception_has_error_code(unsigned int vector)
 			BIT(NP_VECTOR) | BIT(SS_VECTOR) | BIT(GP_VECTOR) |
 			BIT(PF_VECTOR) | BIT(AC_VECTOR);
 
-	return (1U << vector) & exception_has_error_code;
+	return vector < 32 && (1U << vector) & exception_has_error_code;
 }
 
 static inline bool mmu_is_nested(struct kvm_vcpu *vcpu)

From patchwork Fri Oct 12 18:04:29 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10639083
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id CBD29112B
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 12 Oct 2018 18:04:51 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id B97BA2C276
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 12 Oct 2018 18:04:51 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id A98342C294; Fri, 12 Oct 2018 18:04:51 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 55BB32C276
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 12 Oct 2018 18:04:51 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726758AbeJMBib (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 12 Oct 2018 21:38:31 -0400
Received: from mail-vk1-f202.google.com ([209.85.221.202]:37598 "EHLO
        mail-vk1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726695AbeJMBib (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 12 Oct 2018 21:38:31 -0400
Received: by mail-vk1-f202.google.com with SMTP id w65-v6so606906vkw.4
        for <kvm@vger.kernel.org>; Fri, 12 Oct 2018 11:04:49 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=sWH55Fu+zAJ7S5wwD6S9lAC8RCb7zqHkswIT664NlZw=;
        b=asNmZbF9mg0oTS/Njnq/lAFzajwxP8alu4eDC+QivgQ4482nZyFlFegLp0jAHL7qCW
         qFm8wFwq2H5LGz/c1yv9KT8/X0J2KN/17q2RJat9K8xRTseaZKQb93mB8/fEHs47G1qb
         fB2nWVMV1ZlPJaDFXpW4wUDVrIYA1dKwAf5JkaDgDEgxXDoDDOwG3hEy/9JEl4cDkYDn
         bYQg3RvOwjIeNqXjTfvbe88fVfe8RobczSwgmbVCfxzuZrZm0FDHigeYemGzc2Fq7OJM
         h2O4D/m4sC++vY1tDQ0RQWqePtOjx7+U/Jkdyj1sCxfJ35nr0D67hIIkwwCqUBrPLFI5
         Pn9w==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=sWH55Fu+zAJ7S5wwD6S9lAC8RCb7zqHkswIT664NlZw=;
        b=ZjWS8B9Sd5RYGH1dr0GO+Hq4k3rdS7py8bKjOLejX79gpjg5Q0SaBfw2pNuyq3n3g5
         gEv9NvNIZhAP13J0tx1y6BhdL7uOvU4Eu4IVTuz/V/6IKYWOraCaYroYekT1tBONKCpg
         mweB0UKSilwKOG3g4Pmlnd1BL9V3yXDwI/8aPT2r3B1d48IubfRBdM8towNVox2zdNYd
         IVXTIBj+46klk1W1JbylE9JYFMDudmHJ6Prh0IV18QvS9m4uCzmY8Ycyxrd0eY/aph9U
         Hzxswr2bNdOlLeVT/zlvzRkwVbUangnugGU0lu9t8gVXeYSuNf77PihQX82HBT+jERfn
         szCA==
X-Gm-Message-State: ABuFfojUjG+GXcm2XGbDvlVtyRnR8BsWHyEjKIuV7HC0kf2kkuFz7vZU
        Fp5Sn38Q4IsVbHV9erGR4Vrw7quSfCLZvuO+kEqK6FCoWJcLtOWk9lAeSetFi7j123GozGRDRso
        Td8OwJKRPXyWqvwhd/EJ7p3opzCWz5qHzBnpf98bhYTfZt/iNEw9iTDRA1UHVn/k=
X-Google-Smtp-Source: 
 ACcGV60d4Le04Mh2DxS4ys7Ojs7H10Iczfl6MC2UrenbgkfcpIVG1155hViKQMuQZNBirU0VVI95WEVHhZ/ngg==
X-Received: by 2002:a67:8381:: with SMTP id f123mr6935074vsd.27.1539367489236;
 Fri, 12 Oct 2018 11:04:49 -0700 (PDT)
Date: Fri, 12 Oct 2018 11:04:29 -0700
In-Reply-To: <20181012180429.259916-1-jmattson@google.com>
Message-Id: <20181012180429.259916-2-jmattson@google.com>
Mime-Version: 1.0
References: <20181012180429.259916-1-jmattson@google.com>
X-Mailer: git-send-email 2.19.1.331.ge82ca0e54c-goog
Subject: [PATCH 2/2] kvm: x86: Disallow invalid exceptions from userspace
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Userspace should not be able to inject an invalid exception into a kvm
guest via KVM_SET_VCPU_EVENTS, nor should it be able to convert a
valid pending exception to an invalid one via KVM_SET_SREGS.

In particular, only certain exceptions deliver an error code in
protected mode, and no exception delivers an error code in
real-address mode.

Reported-by: syzbot <syzkaller@googlegroups.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
---
 arch/x86/kvm/x86.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 4ab569171ad1..e060c8254cc3 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -3423,6 +3423,14 @@ static int kvm_vcpu_ioctl_x86_set_vcpu_events(struct kvm_vcpu *vcpu,
 	    vcpu->arch.mp_state == KVM_MP_STATE_INIT_RECEIVED)
 		return -EINVAL;
 
+	if (events->exception.injected) {
+		bool has_error_code = is_protmode(vcpu) &&
+			x86_exception_has_error_code(events->exception.nr);
+
+		if (!!events->exception.has_error_code != has_error_code)
+			return -EINVAL;
+	}
+
 	process_nmi(vcpu);
 	vcpu->arch.exception.injected = false;
 	vcpu->arch.exception.pending = events->exception.injected;
@@ -8170,6 +8178,14 @@ static int kvm_valid_sregs(struct kvm_vcpu *vcpu, struct kvm_sregs *sregs)
 			return -EINVAL;
 	}
 
+	if (vcpu->arch.exception.injected || vcpu->arch.exception.pending) {
+		bool has_error_code = (sregs->cr0 & X86_CR0_PE) &&
+			x86_exception_has_error_code(vcpu->arch.exception.nr);
+
+		if (has_error_code != vcpu->arch.exception.has_error_code)
+			return -EINVAL;
+	}
+
 	return 0;
 }
 
